### PR TITLE
Multi-Language Support

### PR DESCRIPTION
**Pull request by FantomWolf182**
Basically the bot can now chat in any language that is on Google Translate. 

The "preferred_language" property accepts a language code or name. Since it's using Google Translate names need to be formatted like on Google Translate. For example if you would like to use Chinese, "preferred_language" would have to be set to "Chinese (Simplified)" or "Chinese (Traditional)" (not case sensitive). Anything the user chat's to the bot is automatically translated to English as I didn't want the AI to bug out. The AI response is then translated to the preferred language the user stated. In case the translation fails, it responds in English. 

The command !setPreferredLanguage can be used in chat, and it takes one argument which is the language you want to switch to.

### Some examples:

Screenshot 2024-08-25 160848
image
image

### CODE REVIEW

**Comment by FantomWolf182** [comment]
There are still some changes that need to be made. 

Mainly if the preferred language is set to English, it shouldn't translate because the ai responds in English anyways.

File: settings.js [hunk 1]
```diff
@@ -19,4 +20,5 @@ export default
     "max_commands": -1, // max number of commands to use in a response. -1 for no limit
     "verbose_commands": true, // show full command syntax
     "narrate_behavior": true, // chat simple automatic actions ('Picking up item!')
-}
\ No newline at end of file
+    "preferred_language": "english", // the bot will respond/message in this language. Secondly all language names are based on google translate's names.
```
Comment (MaxRobinsonTheGreat) [inline settings.js:23]: can you just call it 'language'?

File: settings.js [hunk 2]
```diff
@@ -1,12 +1,13 @@
 export default 
 {
     "minecraft_version": "1.20.4", // supports up to 1.20.4
-    "host": "127.0.0.1", // or "localhost", "your.ip.address.here"
+    "host": "localhost", // or "localhost", "your.ip.address.here"
```
Comment (MaxRobinsonTheGreat) [inline settings.js:4]: remove all changes to settings.js besides language

File: src/agent/agent.js [hunk 3]
```diff
@@ -48,16 +49,18 @@ export class Agent {
...
             ];
             const eventname = settings.profiles.length > 1 ? 'whisper' : 'chat';
-            this.bot.on(eventname, (username, message) => {
+            this.bot.on(eventname, async (username, message) => {
                 if (username === this.name) return;
                 
                 if (ignore_messages.some((m) => message.startsWith(m))) return;
 
-                console.log('received message from', username, ':', message);
+                var translation = await this.handleTranslation(message, "en");
```
Comment (MaxRobinsonTheGreat) [inline src/agent/agent.js:57]: use `let` instead of `var` everywhere

File: src/agent/modes.js [hunk 4]
```diff
@@ -2,10 +2,38 @@ import * as skills from './library/skills.js';
 import * as world from './library/world.js';
 import * as mc from '../utils/mcdata.js';
 import settings from '../../settings.js'
+import translate from 'google-translate-api-x';
 
-function say(agent, message) {
+
+async function handleTranslation(message, lang) {
```
Comment (MaxRobinsonTheGreat) [inline src/agent/modes.js:8]: this is a duplicate function, it should reuse a function from a shared translator file

File: src/agent/agent.js [hunk 5]
```diff
@@ -77,18 +80,46 @@ export class Agent {
...
                 this.bot.emit('finished_executing');
+
             }
 
             this.startEvents();
         });
     }
 
-    cleanChat(message) {
+    async handleTranslation(message, lang) {
```
Comment (MaxRobinsonTheGreat) [inline src/agent/agent.js:94]: handleTranslation should be in its own file, `translator.js` maybe, and be imported where needed.
You shouldn't have to pass prefered_language to it, it should know by reading settings what language to translate to. maybe have a separate function to translate specifically to english

File: src/agent/agent.js [hunk 6]
```diff
@@ -77,18 +80,46 @@ export class Agent {
                 this.handleMessage('system', init_message, 2);
             }
             else {
-                this.bot.chat('Hello world! I am ' + this.name);
+		const translation = await this.handleTranslation("Hello world! I am", `${preferred_lang}`);
+		print(translation)
```
Comment (MaxRobinsonTheGreat) [inline src/agent/agent.js:84]: this should be removed (print doesn't do anything)

File: src/agent/commands/actions.js [hunk 7]
```diff
@@ -286,4 +289,38 @@ export const actionsList = [
             return 'Set npc goal: ' + agent.npc.data.curr_goal.name;
         }
     },
+    {
+    name: '!setPreferredLanguage',
```
Comment (MaxRobinsonTheGreat) [inline src/agent/commands/actions.js:293]: I don't like writing to the settings file at runtime. This should just change the language for the current session

**Review by MaxRobinsonTheGreat — CHANGES_REQUESTED** [review]
got some work for ya!

**Comment by FantomWolf182** [comment]
Ok so a quick update: 

Most of the changes requested have been done. There is a translator.js file which contains everything needed to translate messages, and it's functions are called from the file where necessary. Currently the only thing that has not returned is the in-game command. This is because I can't find a straightforward way in js to update the variable in translator.js from action in actions.js

**Comment by FantomWolf182** [comment]
It's been a while but I might take a shot at this again

**Comment by MaxRobinsonTheGreat** [comment]
@FantomWolf182 Hey thanks for doing all this, it is looking much better. I will test within a week.

Is the google translate api free? I can't imagine it is. Do you have to set up something to get access or does this just work?

**Comment by FantomWolf182** [comment]
yeah its completely free and just works. 

Here is the repo for it: https://github.com/AidanWelch/google-translate-api

**Comment by FantomWolf182** [comment]
The only thing missing right now is the command in game to change the preferred language. However, I wasn't able to rewrite that code to meet your revision, so I'm going to leave it out for now. Everything else should work perfectly.

**Comment by MaxRobinsonTheGreat** [comment]
This was touched up and merged in as #205. Thank you very much @FantomWolf182, this is super useful!

**Comment by FantomWolf182** [comment]
No problem was fun to work on. Thanks for taking the time to review and merge it!

**Pull request merged**